gui/widgets/grid.cpp

```diff
@@ -0,0 +1,640 @@
+}
+
+void GridItemWidget::update() {
+	if (_activeEntry) {
+		updateThumb();
+		setTooltip(_activeEntry->title);
+		markAsDirty();
+	}
+}
+
+void GridItemWidget::move(int x, int y) {
+	Widget::setPos(getRelX() + x, getRelY() + y);
+}
+
+void GridItemWidget::drawWidget() {
+	int thumbHeight = _grid->getThumbnailHeight();
+	int thumbWidth = _grid->getThumbnailWidth();
+	Array<U32String> titleLines;
+	g_gui.getFont().wordWrapText(_activeEntry->title, thumbWidth, titleLines);
+
+	if ((isHighlighted) || (_grid->getSelected() == _activeEntry->entryID)) {
+		// Draw a highlighted BG on hover
+		g_gui.theme()->drawWidgetBackground(Common::Rect(_x - (_grid->_gridXSpacing / 3), _y - (_grid->_gridYSpacing / 3),
+											_x + _w + (_grid->_gridXSpacing / 3), _y + _h + (_grid->_gridYSpacing / 3)), 
+										ThemeEngine::WidgetBackground::kGridItemHighlight);
+	} else {
+		// Draw a BG of the same color as grid area
+		// when it is not highlighted to cover up 
+		// the highlight shadow
+		// FIXME: Find a way to redraw the area around the widget
+		//		 instead of just drawing a cover-up
+		g_gui.theme()->drawWidgetBackground(Common::Rect(_x - 2 * (_grid->_gridXSpacing / 3), _y - 2 * (_grid->_gridYSpacing / 3),
+											_x + _w + 2 * (_grid->_gridXSpacing / 3), _y + _h + 2 * (_grid->_gridYSpacing / 3)), 
+										ThemeEngine::WidgetBackground::kGridItemBackground);
+	}
+
+	// Draw Thumbnail Background
+	g_gui.theme()->drawWidgetBackground(Common::Rect(_x, _y, _x + _grid->getThumbnailWidth(), _y + thumbHeight), 
+										ThemeEngine::WidgetBackground::kThumbnailBackground);
+	
+	// Draw Thumbnail
+	if (_thumbGfx.empty()) {
+		// Draw Title when thumbnail is missing
+		int linesInThumb = MIN(thumbHeight / kLineHeight, (int)titleLines.size());
+		for (int i = 0; i < linesInThumb; ++i) {
+			g_gui.theme()->drawText(Common::Rect(_x, _y + ((thumbHeight - (linesInThumb - 2 * i) * kLineHeight)) / 2,
+							_x + thumbWidth, _y + ((thumbHeight - (linesInThumb - 2 * i) * kLineHeight)) / 2 + kLineHeight),
+							titleLines[i], GUI::ThemeEngine::kStateEnabled, Graphics::kTextAlignCenter,
+							ThemeEngine::kTextInversionNone, 0, true, ThemeEngine::kFontStyleNormal,
+							ThemeEngine::kFontColorAlternate, false);
+		}
+	} else
+		g_gui.theme()->drawSurface(Common::Point(_x, _y), _thumbGfx, true);
+
+	// Draw Platform Icon
+	const Graphics::ManagedSurface *platGfx = _grid->platformToSurface(_activeEntry->platform);
+	if (platGfx) {
+		g_gui.theme()->drawSurface(Common::Point(_x + thumbWidth - 32, _y + thumbHeight - 32), 
+									*platGfx, true);
+	}
+
+	// Draw Flag
+	const Graphics::ManagedSurface *flagGfx = _grid->languageToSurface(_activeEntry->language);
+	if (flagGfx)
+		g_gui.theme()->drawSurface(Common::Point(_x + thumbWidth - flagGfx -> w - 5, _y + 5), 
+									*flagGfx, true);
+
+	// Draw Title
+	if (_grid->_isTitlesVisible) {
+		if (titleLines.size() < 2) {
+			titleLines.push_back(U32String());
+		} else if (titleLines.size() > 2) {
```

Comment:
What about == 2?

Reply:
I didn't want to do anything when titleLines.size() == 2 , since I neither need to add an empty line as with == 1, nor do I need to display ellipses to indicate the title is bigger than 2 lines in > 2.